### PR TITLE
feat(jiva-csi): Add JivaVolumePolicy schema changes

### PR DESCRIPTION
**Pull request by shubham14bajpai**
Signed-off-by: shubham shubham.bajpai@mayadata.io
This PR adds changes corresponding to the schema changes made to JivaVolume and JivaVolumePolicy CRDs. This PR also updates the target path in the JivaVolume spec. The policy name for the volume will be passed as an annotation to the JivaVolume CR.
Refer: https://github.com/openebs/jiva-operator/pull/8

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -25,7 +25,7 @@ install:
   - make bootstrap
 
 before_install:
-  - git clone https://www.github.com/openebs/jiva-operator $GOPATH/src/github.com/openebs/jiva-operator
+  - git clone --branch=policy https://github.com/shubham14bajpai/jiva-operator $GOPATH/src/github.com/openebs/jiva-operator
```
Comment (utkarshmani1997) [inline .travis.yml]: Please remove this once jiva-operator is merged

File: go.mod [hunk 2]
```diff
@@ -24,6 +24,7 @@ require (
 )
 
 replace (
+	github.com/openebs/jiva-operator => github.com/shubham14bajpai/jiva-operator v0.0.0-20200117092318-399d844fdbbd
```
Comment (utkarshmani1997) [inline go.mod]: same comment as above

File: pkg/driver/node.go [hunk 3]
```diff
@@ -465,6 +465,17 @@ func (ns *node) NodePublishVolume(
 		return nil, err
 	}
 
+	// JivaVolume CR may be updated by jiva-operator
+	instance, err := ns.client.GetJivaVolume(req.VolumeId)
```
Comment (utkarshmani1997) [inline pkg/driver/node.go]: use doesVolumeExist function.

File: pkg/kubernetes/client/client.go [hunk 4]
```diff
@@ -127,13 +127,16 @@ func getDefaultLabels(pv string) map[string]string {
...
-	rf := req.GetParameters()["replicaCount"]
+	policyName := req.GetParameters()["policy"]
 	ns, ok := req.GetParameters()["namespace"]
 	if !ok {
 		ns = defaultNS
 	}
-
+	policy := &jv.JivaVolumePolicy{}
+	err := cl.client.Get(context.TODO(), types.NamespacedName{Name: policyName, Namespace: defaultNS}, policy)
+	if err != nil && !errors.IsNotFound(err) {
```
Comment (utkarshmani1997) [inline pkg/kubernetes/client/client.go]: should we create default policies in this case ?

File: pkg/kubernetes/client/client.go [hunk 5]
```diff
@@ -150,43 +153,58 @@ func (cl *Client) CreateJivaVolume(req *csi.CreateVolumeRequest) error {
...
-					},
-				}
-			}(req),
-			ReplicationFactor: func(rf string) string {
-				if rf == "" {
-					return defaultReplicaCount
-				}
-				return rf
-			}(rf),
+			Policy: &jv.JivaVolumePolicySpec{
```
Comment (utkarshmani1997) [inline pkg/kubernetes/client/client.go]: Should we move this to another function ?

File: pkg/kubernetes/client/client.go [hunk 6]
```diff
@@ -150,43 +153,58 @@ func (cl *Client) CreateJivaVolume(req *csi.CreateVolumeRequest) error {
...
+								},
+								Limits: v1.ResourceList{
+									v1.ResourceCPU:    resource.MustParse("0"),
+									v1.ResourceMemory: resource.MustParse("0"),
+								},
+							}
+						}(policy.Spec.Replica.Resources),
+					}},
+
+				Target: jv.TargetSpec{
```
Comment (utkarshmani1997) [inline pkg/kubernetes/client/client.go]: same comment as above

**Review by utkarshmani1997 — CHANGES_REQUESTED** [review]
@shubham14bajpai i have given few minor comments, let's discuss on the same.

File: ci/ci.sh [hunk 7]
```diff
@@ -9,7 +9,7 @@ function initializeTestEnv() {
 	cat <<EOT >> /tmp/parameters.json
 {
         "cas-type": "jiva",
-        "replicaCount": "1"
+        "policy": "example-jivavolumepolicy"
```
Comment (utkarshmani1997) [inline ci/ci.sh:14]: Please add one more test case to test the same without policy

File: pkg/jivavolume/jivavolume.go [hunk 8]
```diff
@@ -21,6 +21,54 @@ import (
...
+	"k8s.io/apimachinery/pkg/api/resource"
+)
+
+const (
+	defaultReplicaSC = "openebs-hostpath"
+)
+
+var (
+	zero              int64
+	defaultTargetSpec = jv.TargetSpec{
```
Comment (utkarshmani1997) [inline pkg/jivavolume/jivavolume.go]: move this to another function `getDefaultTargetSpec`

File: pkg/jivavolume/jivavolume.go [hunk 9]
```diff
@@ -21,6 +21,54 @@ import (
...
+					corev1.ResourceMemory: resource.MustParse("0"),
+				},
+			},
+			Tolerations:       []corev1.Toleration{},
+			NodeSelector:      nil,
+			PriorityClassName: "",
+			Affinity:          &corev1.Affinity{},
+		},
+	}
+	defaultReplicaSpec = jv.ReplicaSpec{
```
Comment (utkarshmani1997) [inline pkg/jivavolume/jivavolume.go]: same comment as above

File: pkg/jivavolume/jivavolume.go [hunk 10]
```diff
@@ -102,16 +150,89 @@ func HasResourceParameters(req *csi.CreateVolumeRequest) ResourceParameters {
...
+	if target.Affinity == nil {
+		target.Affinity = defaultTargetSpec.Affinity
+	}
+	if target.NodeSelector == nil {
+		target.NodeSelector = defaultTargetSpec.NodeSelector
+	}
+	if target.PriorityClassName == "" {
+		target.PriorityClassName = defaultTargetSpec.PriorityClassName
+	}
+	j.jvObj.Spec.Policy.Target = target
```
Comment (utkarshmani1997) [inline pkg/jivavolume/jivavolume.go]: Optional: can we mv entire if cases to a map and then iterate over it to set default values ?

File: pkg/jivavolume/jivavolume.go [hunk 11]
```diff
@@ -102,16 +150,89 @@ func HasResourceParameters(req *csi.CreateVolumeRequest) ResourceParameters {
...
+	if replica.Affinity == nil {
+		replica.Affinity = defaultReplicaSpec.Affinity
+	}
+	if replica.NodeSelector == nil {
+		replica.NodeSelector = defaultReplicaSpec.NodeSelector
+	}
+	if replica.PriorityClassName == "" {
+		replica.PriorityClassName = defaultReplicaSpec.PriorityClassName
+	}
+	j.jvObj.Spec.Policy.Replica = replica
```
Comment (utkarshmani1997) [inline pkg/jivavolume/jivavolume.go]: same comment as above

**Review by utkarshmani1997 — CHANGES_REQUESTED** [review]
@shubham14bajpai i have provided few minor comments, please have a look at it. I will approve this once jiva-operator PR is merged.

File: ci/ci.sh (hunk 7 above)
Comment (shubham14bajpai) [reply ci/ci.sh:14]: Without the policy, the default replicationfactor will be 3 which cannot be tested on a single node.

**Review by utkarshmani1997 — APPROVED** [review]
LGTM

File: go.mod [hunk 12]
```diff
@@ -24,6 +24,7 @@ require (
 )
 
 replace (
+	github.com/openebs/jiva-operator => github.com/shubham14bajpai/jiva-operator a3a1919cdf152e0653a547b084f831971a22fd1f
```
Comment (prateekpandey14) [inline go.mod]: we have to remove this change ? once the operator changes merged

**Review by prateekpandey14 — APPROVED** [review]
some minor not picks and cleanups  deps , otherwise LGTM

**Pull request merged**